include/nbl/video/alloc/CSingleBufferSubAllocator.h

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_VIDEO_C_SINGLE_BUFFER_SUB_ALLOCATOR_H_
+#define _NBL_VIDEO_C_SINGLE_BUFFER_SUB_ALLOCATOR_H_
+
+#include "nbl/video/alloc/IBufferAllocator.h"
+
+#include <type_traits>
+
+namespace nbl::video
+{
+
+// address allocator gives offsets
+// reserved allocator allocates memory to keep the address allocator state inside
+template<class AddrAllocator, class ReservAllocator=core::allocator<uint8_t>>
+class CSingleBufferSubAllocator : public IBufferAllocator
+{
+  public:
+        using AddressAllocator = AddrAllocator;
+        using ReservedAllocator = ReservAllocator;
+        using size_type = AddressAllocator::size_type;
+        using value_type = AddressAllocator::size_type;
+        static constexpr value_type invalid_value = AddressAllocator::invalid_address;
+
+        // constructors
+        template<typename... Args>
+        inline CSingleBufferSubAllocator(asset::SBufferRange<IGPUBuffer>&& _bufferRange, ReservedAllocator&& _reservedAllocator, const value_type maxAllocatableAlignment, Args&&... args) :
+            m_addressAllocator(
+                _reservedAllocator.allocate(AddressAllocator::reserved_size(maxAllocatableAlignment,static_cast<size_type>(_bufferRange.size), args...), _NBL_SIMD_ALIGNMENT),
+                static_cast<size_type>(_bufferRange.offset), 0u, maxAllocatableAlignment, static_cast<size_type>(_bufferRange.size), std::forward<Args>(args)...
+            ), m_reservedAllocator(std::move(_reservedAllocator)), m_buffer(std::move(_bufferRange.buffer))
+        {
+            assert(_bufferRange.isValid());
```

Comment:
you std::move the buffer in the constructor initializer list so bufferRange.buffer would be nullptr and this assert would always hit

Reply:
ok, then a different assert is needed which essentially does the same thing